packages/foundation/src/next/composed.tsx

```diff
@@ -42,8 +43,96 @@ export function composed<
+ * Imposes a separation of concern and centralizes styling processing to increase ease of use and robustness
+ * in how components use and apply styling and theming.
+ *
+ * Automatically merges and applies themes and styles with theme / styleprops having the highest priority.
+ * State component, if provided, is passed in props for processing. Props from state / user are automatically processed
+ * and styled before finally being passed to view.
+ *
+ * State components should contain all stateful behavior and should not generate any JSX, but rather simply call the view prop.
+ * Views should simply be stateless pure functions that receive all props needed for rendering their output.
+ * State component is optional. If state is not provided, created component is essentially a functional stateless component.
+ *
+ * @param baseComponent - base component to recompose
+ * @param options - component Component options. See IComponentOptions for more detail.
+ */
+export function composed<
+  TComponentProps extends ValidProps & ISlottableProps<TComponentSlots>,
+  TTokens,
+  TStyleSet extends IStyleSet<TStyleSet>,
+  TViewProps extends TComponentProps = TComponentProps,
+  TComponentSlots = {},
+  TStatics = {}
+>(
+  baseComponent: React.FunctionComponent,
+  options: IRecompositionComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
+): React.FunctionComponent<TComponentProps> & TStatics;
+
+export function composed<
+  TComponentProps extends ValidProps & ISlottableProps<TComponentSlots>,
+  TTokens,
+  TStyleSet extends IStyleSet<TStyleSet>,
+  TViewProps extends TComponentProps = TComponentProps,
+  TComponentSlots = {},
+  TStatics = {}
+>(
+  baseComponentOrOptions:
+    | React.FunctionComponent
+    | IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics> = {},
+  recompositionOptions?: IRecompositionComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
 ): React.FunctionComponent<TComponentProps> & TStatics {
+  const baseComponentOptions = (baseComponentOrOptions as ISlotCreator<
+    TComponentProps,
+    any,
+    TTokens,
+    TStyleSet,
+    TViewProps,
+    TComponentSlots,
+    TStatics
+  >).__options;
+
+  // Check if we are composing or recomposing.
+  let options: IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>;
+  if (typeof baseComponentOrOptions === 'function' && baseComponentOptions) {
+    let slots: ISlotComponent<TComponentProps, TComponentSlots>;
+    const baseComponentSlots = baseComponentOptions.slots ? baseComponentOptions.slots : ({} as ISlotDefinition<Required<TComponentSlots>>);
+    const recompositionSlots = recompositionOptions ? (recompositionOptions.slots ? recompositionOptions.slots : {}) : {};
+
+    if (typeof baseComponentSlots === 'function') {
```

Comment:
Why would `baseComponentSlots` be a function? It seems like you'd not want slots to be a function of props.

Reply:
What happens for something like `Button` where the default `root` slot can change from `button` to `a` when `href` is set?

Reply:
```
const resolveWith = (thing, data) => typeof thing === 'function' ? thing(data) : thing;

slots = props => {
  ...resolveWith(baseComponentSlots, props),
  ...resolveWith(recompositionSlots, props)
};
```